std/algorithm/iteration.d

```diff
@@ -4035,16 +4038,97 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
 }
 
 // Pairwise summation http://en.wikipedia.org/wiki/Pairwise_summation
-private auto sumPairwise(Result, R)(R r)
+private auto sumPairwise(F, R)(R data)
+if (isInputRange!R && !isInfinite!R)
 {
-    static assert (isFloatingPoint!Result);
-    switch (r.length)
+    import core.bitop : bsf;
+    // Works for r with at least length < 2^^(64 + log2(16)), in keeping with the use of size_t
+    // elsewhere in std.algorithm and std.range on 64 bit platforms. The 16 in log2(16) comes
+    // from the manual unrolling in sumPairWise16
+    F[64] store = void;
+    size_t idx = 0;
+
+    auto collapseStore(T)(T k)
```

Comment:
@andralex I had to make this a template because `size_t` is no good as a generic index/length type for ranges (e.g. sdt.range.iota of ulong on 32bit). I now don't know what to do with e.g. line 4087 which uses size_t as counter. Should I use `ulong`? What is the maximum allowed length for a range?